2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+nuget.exe.
+This module will only use REST APIs to get and publish nupkgs.
+This module would be shipped in PowerShell 7.
+
+### Side-by-side with PowerShellGet
+
+To ease transition due to the enormity of the breaking change, this module will
+have a different names for the cmdlets to allow for side by side use
+with the existing PowerShellGet module.
+Since the current PowerShellGet 2.x version is a script module and this new one
+is a C# based module, they can coexist side-by-side.
+
+### Cmdlet compatibility
+
+Function wrappers will be provided to provide compatibility with the two most often
+used cmdlets: `Install-Module` and `Find-Module`.
+These cmdlets will output a warning to use the new cmdlets.
+
+### Local cache
+
+Rather than connecting to PSGallery and other repositories for finding modules,
+the user must explicitly request the current metadata of
+modules from the repositories and that is cached locally.
+`Find-PSResource` (see below) works against this local cache.
+This will also enable changes in PowerShell to use `Find-PSResource -Type Command` to look
+in this cache when it can't find a command and suggest to the user the module to install to
+get that command.
+This will be a local json file containing only the latest version of each module.
+The cache will be stored in the user path.
+There is no system cache that is shared.
+
+Example cache entry:
+
+```json
+{
+  "name": "This is my module",
+  "version": "1.0.0.0",
+  "type": "module",
+  "tags": [
+    "Linux",
+    "PSEdition_Core",
+    "PSEdition_Desktop",
+    "AzureAutomation"
+  ]
+}
+```
+
+An example cache with 5000 resources is ~700KB in compressed json form.
+
+### Automatic updating of the cache
+
+On any operation that reaches out to a repository, a REST API will be called to
+see if the hash of the cache matches the current cache and if not, a new one
+is downloaded.
+If the repository doesn't support this new API, it falls back to current behavior
+in PSGet v2.
+
+### Repository management
+
+`Register-PSResourceRepository` will allow for registering additional repositories.
+A `-Default` switch enables registering PSGallery should it be accidentally removed.
+The `-URL` will accept the HTTP address without the need to specify `/api/v2` as
+that will be assumed and discovered at runtime.
+Support for local filesystem repositories must be maintained.
+A `-InstallationPolicy` switch accepts `Trusted` or `Untrusted` (default) indicating
+whether to prompt the user when installing resources from that repository.
+
+`Get-PSResourceRepository` will list out the registered repositories.
+
+`Set-PSResourceRepository` can be used to update a repository URL or installation policy.
+
+`Unregister-PSResourceRepository` can be used to un-register a repository.
+
+### Finding resources
+
+Current design of PowerShellGet is to have different cmdlets for each of the
+different resources types that it supports:
+
+`Find-Command`
```

Comment:
Why separate cmdlets instead of just a type parameter if people want to get specific? To a lot of users, they don't think about the difference between command and script. `Get-Command` ignore the difference between command type by default, and has a `-Type` parameter if people want to get specific. I think the same should apply here as well to ease users into PowerShell who don't understand the difference and may never need to.

Reply:
This refers to the current design.  The new design does what you want (I think) by offering a single `Find-PSResource` command with a `-Type` parameter.  My issue with this is that I believe the most common usage is to search for modules.  And we've been training folks to use `Find-Module` ever since v5 shipped.  I'm OK with adding `Find-PSResourc` but I think `Find-Module` needs to remain (and not warn).

Reply:
Ugh (on me), you're right. Glanced over this section too quickly.

Reply:
@rkeithhill current proposal has wrappers for `Install-Module` and `Find-Module`, but with warnings.  Maybe we keep those as cmdlets without warnings.  Would like more input on this.

Reply:
Don't think warnings would be necessary, unless the intent was to slowly move people away from using Install/find-module.  I like the idea of having the option to use either the more generic wrapper or the specific cmd it's wrapping.